src/Database/LSMTree/Internal/RefCount.hs

```diff
@@ -0,0 +1,115 @@
+  , addReference
+  , readRefCount
+  , removeReference
+  , replaceFinaliser
+  , finalise
+  , RefCount (..)
+  ) where
+
+import           Control.DeepSeq
+import           Control.Monad
+import           Control.Monad.Class.MonadThrow
+import           Control.Monad.Primitive
+import           Data.Maybe
+import           Data.Primitive.MutVar
+
+-- | A reference counter with an optional finaliser action. Once the reference
+-- count reaches @0@, the finaliser will be run.
+data RefCounter m = RefCounter {
+    countVar     :: !(MutVar (PrimState m) RefCount)
+  , finaliserVar :: !(MutVar (PrimState m) (Maybe (m ())))
+  }
+
+instance NFData (RefCounter m) where
+  rnf RefCounter{countVar, finaliserVar} =
+      rwhnf countVar `seq` rwhnf finaliserVar
+
+{-# SPECIALISE unsafeMkRefCounterN :: RefCount -> Maybe (IO ()) -> IO (RefCounter IO) #-}
+-- | Like 'mkRefCounterN', but throws an error if the initial @n < 1@
+unsafeMkRefCounterN :: PrimMonad m => RefCount -> Maybe (m ()) -> m (RefCounter m)
+unsafeMkRefCounterN !n finaliser = mkRefCounterN n finaliser >>= \case
+    Nothing -> error "unsafeMkRefCounterN: n < 1"
+    Just rc -> pure rc
+
+{-# SPECIALISE mkRefCounterN :: RefCount -> Maybe (IO ()) -> IO (Maybe (RefCounter IO)) #-}
+-- | Make a reference counter with initial value @n@. An optional finaliser is
+-- run when the reference counter reaches @0@.
+--
+-- Returns 'Nothing' if @n < 1@, and 'Just' otherwise.
+mkRefCounterN :: PrimMonad m => RefCount -> Maybe (m ()) -> m (Maybe (RefCounter m))
+mkRefCounterN !n finaliser
+  | n < RefCount 1 = pure Nothing
+  | otherwise = do
+    -- evaluate the finaliser a little bit before we store it
+    let !() = case finaliser of
+                Nothing       -> ()
+                Just !_action -> ()
+    countVar <- newMutVar n
+    finaliserVar <- newMutVar finaliser
+    pure $! Just $! RefCounter{countVar, finaliserVar}
+
+{-# SPECIALISE mkRefCounter1 :: Maybe (IO ()) -> IO (RefCounter IO) #-}
+-- | Make a reference counter with initial value @1@. An optional finaliser is
+-- run when the reference counter reaches @0@.
+mkRefCounter1 :: PrimMonad m => Maybe (m ()) -> m (RefCounter m)
+mkRefCounter1 finaliser = fromJust <$> mkRefCounterN (RefCount 1) finaliser
+
+data RefCountError = RefCountAlreadyZero
+  deriving stock (Show, Eq)
+  deriving anyclass Exception
+
+{-# SPECIALISE addReference :: RefCounter IO -> IO () #-}
+-- | Increase the reference counter by one.
+--
+-- If the reference counter was already at @0@, a 'RefCountError' error is
+-- thrown.
+addReference :: (PrimMonad m, MonadThrow m) => RefCounter m -> m ()
+addReference RefCounter{countVar} = do
+    success <- atomicModifyMutVar' countVar $ \c ->
+      if c == RefCount 0 then (c, False)
+                         else (incRefCount c, True)
+    unless success $ throwIO RefCountAlreadyZero
+
+{-# SPECIALISE readRefCount :: RefCounter IO -> IO RefCount #-}
+readRefCount :: PrimMonad m => RefCounter m -> m RefCount
```

Comment:
--| Warning: reading the current reference count is inherently racy as there is no way to reliably act on the information. It can be useful for debugging.